### PR TITLE
Added An Overview of Chalk

### PR DESCRIPTION
**Pull request by sunjay**
This is the overview of [chalk](https://github.com/rust-lang-nursery/chalk) given to me by Niko in the #wg-traits Discord channel.

r? @nikomatsakis 

[Rendered](https://github.com/sunjay/rustc-guide/blob/8cfd5c5aede5fe17cc286c230ccd2eba4cc966bd/src/chalk-overview.md)

Even more rendered: (click for full size)

screen shot 2018-05-14 at 14 44 47-fullpage

### CODE REVIEW

File: src/chalk-overview.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
+# An Overview of Chalk
+
+> Chalk is under heavy development, so if any of these links are broken or if
+> any of the information is inconsistent with the code or outdated, please open
+> an issue so we can fix it. If you are able to fix the issue yourself, we would
+> love your contribution!
```
Comment (mark-i-m) [inline src/chalk-overview.md]: Maybe a link to the repo here?

File: src/chalk-overview.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+[Chalk][chalk] recasts Rust's trait system explicitly in terms of logic
+programming by "lowering" Rust code into a kind of logic program we can then
```
Comment (mark-i-m) [inline src/chalk-overview.md]: Perhaps a link to the lowering chapters would be good here?
- https://rust-lang-nursery.github.io/rustc-guide/traits-lowering-to-logic.html
- https://rust-lang-nursery.github.io/rustc-guide/traits-lowering-rules.html

File: src/chalk-overview.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+execute queries against. Its goal is to be an executable, highly readable
+specification of the Rust trait system.[^negativechalk]
```
Comment (mark-i-m) [inline src/chalk-overview.md]: I don't understand the footnote here?

File: src/chalk-overview.md [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+* [Lowering Rust traits to logic](http://smallcultfollowing.com/babysteps/blog/2017/01/26/lowering-rust-traits-to-logic/)
+* [Unification in Chalk, part 1](http://smallcultfollowing.com/babysteps/blog/2017/03/25/unification-in-chalk-part-1/)
+* [Unification in Chalk, part 2](http://smallcultfollowing.com/babysteps/blog/2017/04/23/unification-in-chalk-part-2/)
+* [Negative reasoning in Chalk](http://aturon.github.io/blog/2017/04/24/negative-chalk/)
+* [Query structure in chalk](http://smallcultfollowing.com/babysteps/blog/2017/05/25/query-structure-in-chalk/)
+* [Cyclic queries in chalk](http://smallcultfollowing.com/babysteps/blog/2017/09/12/tabling-handling-cyclic-queries-in-chalk/)
+* [An on-demand SLG solver for chalk](http://smallcultfollowing.com/babysteps/blog/2018/01/31/an-on-demand-slg-solver-for-chalk/)
+
+## Parsing
```
Comment (mark-i-m) [inline src/chalk-overview.md]: Perhaps add a note that this is completely orthogonal to the Rust AST and grammar.

File: src/chalk-overview.md [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+Chalk is designed to be incorporated with the Rust compiler, so the syntax and
+concepts it deals with heavily borrow from Rust. It is convenient for the sake
+of testing to be able to run chalk on its own, so chalk includes a parser for a
+Rust-like syntax.
+
+The parser takes that syntax and produces an [Abstract Syntax Tree (AST)][ast].
+You can find the [complete definition of the AST][chalk-ast] in the source code.
+
+The syntax contains things from Rust that we know and love for example traits,
```
Comment (mark-i-m) [inline src/chalk-overview.md]: Comma needed after "love"

File: src/chalk-overview.md [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+impls, and struct definitions. Parsing is often the first "phase" of
+transformation that a program goes through in order to become a format that
+chalk can understand.
+
+## Lowering
+
+After parsing, there is a "lowering" phase. This aims to convert traits/impls
+into "program clauses". A [`ProgramClause` (source code)][programclause] is
```
Comment (mark-i-m) [inline src/chalk-overview.md]: Perhaps a link to https://rust-lang-nursery.github.io/rustc-guide/traits-goals-and-clauses.html would also be useful here?

File: src/chalk-overview.md [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+essentially one of the following:
+
+* A [clause] of the form `consequence :- conditions` where `:-` is read as
+  "if" and `conditions = cond1 && cond2 && ...`
+* A universally quantified clause of the form
+  `forall<T> { consequence :- conditions }`
+  * `forall<T> { ... }` is used to represent [universal quantification]. See the
```
Comment (mark-i-m) [inline src/chalk-overview.md]: Perhaps "universal quantification" would be a good term to add to the glossary?

File: src/chalk-overview.md [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+    section on [Lowering to logic][lowering-forall] for more information.
+  * A key thing to note about `forall` is that we don't allow you to "quantify"
+    over traits, only types and regions (lifetimes). That is, you can't make a
+    rule like `forall<Trait> { u32: Trait }` which would say "`u32` implements
+    all traits". You can however say `forall<T> { T: Trait }` meaning "`Trait`
+    is implemented by all types".
```
Comment (mark-i-m) [inline src/chalk-overview.md]: :+1: This is a great example

File: src/chalk-overview.md [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+  * `forall<T> { ... }` is represented in the code using the [`Binders<T>`
+    struct][binders-struct].
+
+This is the phase where we encode the rules of the trait system into logic. For
```
Comment (mark-i-m) [inline src/chalk-overview.md]: s/This/Lowering/

File: src/chalk-overview.md [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+example, if we have:
```
Comment (mark-i-m) [inline src/chalk-overview.md]: "if we have" -> "if we have the following Rust"

File: src/chalk-overview.md [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+
+```rust,ignore
+impl<T: Clone> Clone for Vec<T> {}
+```
+
+We generate:
```
Comment (mark-i-m) [inline src/chalk-overview.md]: "We generate" -> "We generate the following program clause"

File: src/chalk-overview.md [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+This rule dictates that `Vec<T>: Clone` is only satisfied if `T: Clone` is also
+satisfied (i.e. "provable").
+
+### Well-formedness checks
+
+As part of lowering from the AST to the internal IR, we also do some "well
+formedness" checks. See the [source code][well-formedness-checks] for where
+those are done. The call to `record_specialization_priorities` checks
+"coherence" which means that it ensures that two impls of the same trait for the
+same type cannot exist.
```
Comment (mark-i-m) [inline src/chalk-overview.md]: I'm guessing that this actually checks that specialization rules are not violated (i.e. overlapping impls are "valid")? In other words, if we wanted to implement specialization, would we just change the implementation of `record_specialization_priorities`?

File: src/chalk-overview.md [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+The second intermediate representation in chalk is called, well, the "ir". :)
+The [IR source code][ir-code] contains the complete definition. The
+`ir::Program` struct contains some "rust things" but indexed and accessible in
+a different way. This is sort of analogous to the [HIR] in Rust.
+
+For example, if you have a type like `Foo<Bar>`, we would represent `Foo` as a
+string in the AST but in `ir::Program`, we use numeric indices (`ItemId`).
+
+In addition to `ir::Program` which has "rust-like things", there is also
+`ir::ProgramEnvironment` which is "pure logic". The main field in that is
```
Comment (mark-i-m) [inline src/chalk-overview.md]: "that" -> "that struct"?

File: src/chalk-overview.md [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+`program_clauses` which contains the `ProgramClause`s that we generated
+previously.
+
+## Rules
+
+The `rules` module works by iterating over every trait, impl, etc. and emitting
+the rules that come from each one. The traits section of the rustc-guide (that
```
Comment (mark-i-m) [inline src/chalk-overview.md]: Perhaps link to https://rust-lang-nursery.github.io/rustc-guide/traits-lowering-rules.html for convenience?

File: src/chalk-overview.md [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+you are currently reading) contains the most up-to-date reference on that.
+
+The `ir::ProgramEnvironment` is created [in this module][rules-environment].
+
+## Testing
+
+TODO: Basically, [there is a macro](https://github.com/rust-lang-nursery/chalk/blob/94a1941a021842a5fcb35cd043145c8faae59f08/src/solve/test.rs#L112-L148)
+that will take syntax and run it through the full pipeline described above.
```
Comment (mark-i-m) [inline src/chalk-overview.md]: Rust or chalk syntax? (I'm guessing the latter)

File: src/chalk-overview.md (hunk 3 above)
Comment (sunjay) [reply src/chalk-overview.md]: This description is a paraphrased (and copied) version of the blog post in the footnote so I thought I should cite my sources :stuck_out_tongue_closed_eyes:

File: src/chalk-overview.md (hunk 8 above)
Comment (sunjay) [reply src/chalk-overview.md]: Thanks! All of this is from @nikomatsakis, I just typed it up and made it look pretty :smile:

File: src/chalk-overview.md (hunk 12 above)
Comment (sunjay) [reply src/chalk-overview.md]: @nikomatsakis can answer this for sure, but my understanding is that this function is named this way because it will eventually help with specialization even though that isn't fully implemented yet. The description about the coherence checks is still correct.

File: src/chalk-overview.md (hunk 7 above)
Comment (sunjay) [reply src/chalk-overview.md]: Maybe? It isn't exactly Rust specific, but we certainly could add it if you think it would be useful. :)

File: src/chalk-overview.md (hunk 6 above)
Comment (sunjay) [reply src/chalk-overview.md]: This is done. There is a "See also" link after the bullets below.

**Comment by sunjay** [comment]
@mark-i-m I believe I have addressed all of your comments. GitHub's UI is weird and it hid some of them from me so it took a few tries to get all of them. Let me know if there are any more that I missed or if you have any more feedback. Thanks for your review!

**Comment by mark-i-m** [comment]
LGTM :) Thanks!

I would be fine with merging this now, but if you want, we can also wait for a review from Niko.

**Comment by sunjay** [comment]
Sure. Let's wait for Niko to see if everything is accurately written. :)

**Comment by nikomatsakis** [comment]
Looks good to me!
